File: nara_service/frontend/src/app/prometheus/types/hierarchyTypes.ts
/**
 * 문서 간 상하위 위계 관계 타입 정의
 */

export enum HierarchyRelationType {
  PARENT_OF = "PARENT_OF",
  CHILD_OF = "CHILD_OF",
  INCLUDES = "INCLUDES",
}

export interface HierarchyType {
  id: string;
  name: string;
  description: string;
  direction: "up" | "down";
  examples: string[];
}

export const HIERARCHY_TYPES: Record<HierarchyRelationType, HierarchyType> = {
  [HierarchyRelationType.PARENT_OF]: {
    id: "PARENT_OF",
    name: "상위 문서",
    description: "이 문서는 대상 문서의 상위 개념/정책입니다",
    direction: "down",
    examples: ["총괄 정책 → 세부 정책", "법률 → 시행령"],
  },
  [HierarchyRelationType.CHILD_OF]: {
    id: "CHILD_OF",
    name: "하위 문서",
    description: "이 문서는 대상 문서의 하위 개념/세부사항입니다",
    direction: "up",
    examples: ["세부 지침 → 상위 정책", "시행규칙 → 법률"],
  },
  [HierarchyRelationType.INCLUDES]: {
    id: "INCLUDES",
    name: "포함",
    description: "이 문서는 대상 문서를 포함합니다",
    direction: "down",
    examples: ["종합 보고서 → 부분 보고서", "전체 계획 → 세부 항목"],
  },
};

export function getHierarchyTypesList(): HierarchyType[] {
  return Object.values(HIERARCHY_TYPES);
}

export function getHierarchyTypeName(typeId: string): string {
  const hierarchyType = HIERARCHY_TYPES[typeId as HierarchyRelationType];
  return hierarchyType ? hierarchyType.name : typeId;
}

export function isHierarchyRelation(edgeType?: string): boolean {
  if (!edgeType) return false;
  return Object.keys(HIERARCHY_TYPES).includes(edgeType);
}

File: nara_service/backend/app/core/hierarchy_types.py
"""
문서 간 상하위 위계 관계 타입 정의
"""

from enum import Enum
from typing import Dict, List


class HierarchyRelationType(str, Enum):
    """위계 관계 타입"""
    PARENT_OF = "PARENT_OF"  # 상위 문서
    CHILD_OF = "CHILD_OF"  # 하위 문서
    INCLUDES = "INCLUDES"  # 포함


# 위계 관계 메타데이터
HIERARCHY_METADATA: Dict[str, Dict] = {
    "PARENT_OF": {
        "name": "상위 문서",
        "description": "이 문서는 대상 문서의 상위 개념/정책입니다",
        "direction": "down",  # 상위에서 하위로
        "inverse": "CHILD_OF",
        "examples": ["총괄 정책 → 세부 정책", "법률 → 시행령"]
    },
    "CHILD_OF": {
        "name": "하위 문서",
        "description": "이 문서는 대상 문서의 하위 개념/세부사항입니다",
        "direction": "up",  # 하위에서 상위로
        "inverse": "PARENT_OF",
        "examples": ["세부 지침 → 상위 정책", "시행규칙 → 법률"]
    },
    "INCLUDES": {
        "name": "포함",
        "description": "이 문서는 대상 문서를 포함합니다",
        "direction": "down",
        "inverse": "INCLUDED_IN",
        "examples": ["종합 보고서 → 부분 보고서", "전체 계획 → 세부 항목"]
    }
}


def get_hierarchy_types() -> List[Dict]:
    """위계 관계 타입 목록 반환 (API 응답용)"""
    return [
        {
            "id": rel_type,
            "name": metadata["name"],
            "description": metadata["description"],
            "direction": metadata["direction"],
            "examples": metadata["examples"]
        }
        for rel_type, metadata in HIERARCHY_METADATA.items()
    ]


def get_inverse_relation(rel_type: str) -> str:
    """역방향 관계 타입 반환"""
    if rel_type in HIERARCHY_METADATA:
        return HIERARCHY_METADATA[rel_type]["inverse"]
    return rel_type
